services/proxy/pkg/middleware/public_share_auth.go

```diff
@@ -5,68 +5,82 @@ import (
 	"strings"
 
 	gateway "github.com/cs3org/go-cs3apis/cs3/gateway/v1beta1"
+	"github.com/owncloud/ocis/v2/ocis-pkg/log"
 )
 
 const (
-	headerRevaAccessToken   = "x-access-token"
+	_headerRevaAccessToken  = "x-access-token"
 	headerShareToken        = "public-token"
 	basicAuthPasswordPrefix = "password|"
 	authenticationType      = "publicshares"
+
+	_paramSignature  = "signature"
+	_paramExpiration = "expiration"
 )
 
-// PublicShareAuth ...
-func PublicShareAuth(opts ...Option) func(next http.Handler) http.Handler {
-	options := newOptions(opts...)
-	logger := options.Logger
+// PublicShareAuthenticator is the authenticator which can authenticate public share requests.
+// It will add the share owner into the request context.
+type PublicShareAuthenticator struct {
+	Logger            log.Logger
+	RevaGatewayClient gateway.GatewayAPIClient
+}
 
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			shareToken := r.Header.Get(headerShareToken)
-			if shareToken == "" {
-				shareToken = r.URL.Query().Get(headerShareToken)
-			}
+func (a PublicShareAuthenticator) Authenticate(r *http.Request) (*http.Request, bool) {
+	if !isPublicPath(r.URL.Path) {
+		return nil, false
+	}
 
-			// Currently we only want to authenticate app open request coming from public shares.
-			if shareToken == "" {
-				// Don't authenticate
-				next.ServeHTTP(w, r)
-				return
-			}
+	query := r.URL.Query()
+	shareToken := r.Header.Get(headerShareToken)
+	if shareToken == "" {
+		shareToken = query.Get(headerShareToken)
+	}
 
-			var sharePassword string
-			if signature := r.URL.Query().Get("signature"); signature != "" {
-				expiration := r.URL.Query().Get("expiration")
-				if expiration == "" {
-					logger.Warn().Str("signature", signature).Msg("cannot do signature auth without the expiration")
-					next.ServeHTTP(w, r)
-					return
-				}
-				sharePassword = strings.Join([]string{"signature", signature, expiration}, "|")
-			} else {
-				// We can ignore the username since it is always set to "public" in public shares.
-				_, password, ok := r.BasicAuth()
+	if shareToken == "" {
+		// If the share token is not set then we don't need to inject the user to
+		// the request context so we can just continue with the request.
+		return r, true
```

Comment:
This confuses me. If any Authenticator returns true it means the request is authenticated.

Wouldn't it mean that any request with no `shareToken` would be considered authenticated from the `PublicShareAuthenticator`?

Reply:
> Wouldn't it mean that any request with no shareToken would be considered authenticated from the PublicShareAuthenticator?

If the request is a normal public share request and doesn't contain a shareToken in the header or query then we just handle it like a normal public share request and forward it to the next service. Most of the time this service is reva which authenticates the request itself.

If we would return false here then you wouldn't be able to do normal public share requests.
This middleware was only introduced for public share requests to services which do require an authenticated context and aren't in reva e.g. the capabilities request in OCS.

Reply:
Mmh. I still don't get it, sorry. As far as I understood, if any authenticator middleware returns true, the whole request is being considered authenticated. 

But this can return true and still other authentication checks are being done? What am I misunderstanding?

Reply:
Ok let me try again.

A request like `PROPFIND https://ocis.example.com/remote.php/dav/public-files/tVeVwRxHnspGzyH` doesn't require an authentication since it is a public share. But for technical reasons Reva always requires a user in the context. For this reason Reva fetches the share owner and puts them in the context.

This means the proxy doesn't need to do anything with the Request except forwarding it to Reva. With the new authenticator this means the authenticator needs to return `true` otherwise the authentication will be seen as failed and the user receives a `HTTP 401`.

If a client wants to request something which is not in Reva but still requires an authenticated context then they can add the `public-token` header or query so that the public share authenticator adds the share owner to the context.

So to sum up what the Public Share Authenticator needs to do is to just "allow the request or add the share owner to the context and allow the request".